zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,180 @@
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultZkNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return endpointList;
+    }
+
+    final class ZkWatcher implements Watcher, StatCallback {
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(zNode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(zNode, true, this, null);
+                }
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                case NOAUTH:
+                    return;
+                default:
+                    // Retry errors
+                    zooKeeper.exists(zNode, true, this, null);
+                    return;
+            }
+
+            byte[] nodeData = null;
+            if (exists) {
+                try {
+                    nodeData = zooKeeper.getData(zNode, false, null);
+                } catch (Exception e) {
+                    logger.warn("error to get zNode data:" + e.getMessage());
+                    return;
+                }
+            }
+            if (nodeData == null && nodeData != prevData ||
+                nodeData != null && !Arrays.equals(prevData, nodeData)) {
+                prevData = nodeData;
+                try {
+                    endpointList = converter.convert(prevData);
+                } catch (IllegalArgumentException exception) {
+                    logger.warn("error to convert zNode value to EndpointGroup:" + exception.getMessage() +
+                                ",invalid value:" + new String(prevData, Charsets.UTF_8));
+                }
+            }
+        }
+    }
+
+    /**
+     *  Closes the underlying Zookeeper connection.
+     * @throws EndpointGroupException Closing exception
+     */
+    public void close() throws EndpointGroupException {
+        try {
+            zooKeeper.close();
+        } catch (InterruptedException e) {
+            throw new EndpointGroupException(
+                    "error to close underlying ZooKeeper connection :" + e.getMessage());
```

Comment:
- error to -> failed to
- Remove the space before colon (:) and the colon
- `new EndpointGroupException("failed to close ... connection", e);`